### PR TITLE
Remove the distinction between query and filter context in QueryBuilders

### LINKED ISSUE #35293: Function score functions filter executed in query context
**Elasticsearch version** (`bin/elasticsearch --version`):

Version: 6.4.2, Build: default/deb/04711c2/2018-09-26T13:34:09.098244Z, JVM: 1.8.0_191

**Description of the problem including expected versus actual behavior**:

When executing function score query with a bunch of function, each with its `filter` clause I would expect them (the `filter` queries) to be executed in [filter context](https://www.elastic.co/guide/en/elasticsearch/reference/6.4/query-filter-context.html), but they seem to be executed in query context instead.

**Steps to reproduce**:

### Step 1: Index 2 documents:

```
curl -H 'Content-Type: application/x-ndjson' -XPOST localhost:9200/_bulk --data-binary '
{"index":{"_index":"test","_type":"_doc","_id":"1"}}
{"tags":["a", "b"]}
{"index":{"_index":"test","_type":"_doc","_id":"2"}}
{"tags":["a", "c"]}
'
```

### Step 2: Prepare query

```json
{
  "query": {
    "function_score": {
      "functions": [
	{
	  "filter": {
	    "bool": {
	      "must":   {"match": {"tags": "a"}},
	      "should": {"match": {"tags": "c"}}
	    }
	  },
	  "weight": 100
	}
      ]
    }
  }
}
```

### Step 3: Observe the results

```json
{
  "took": 21,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": 2,
    "max_score": 100,
    "hits": [
      {
        "_index": "test",
        "_type": "_doc",
        "_id": "2",
        "_score": 100,
        "_source": {
          "tags": [
            "a",
            "c"
          ]
        }
      },
      {
        "_index": "test",
        "_type": "_doc",
        "_id": "1",
        "_score": 100,
        "_source": {
          "tags": [
            "a",
            "b"
          ]
        }
      }
    ]
  }
}
```

### Analysis

Both documents get 100 score, which means that the (only) function from `function_score` applied there. Which means that the `bool` query from its `filter` clause matched both documents. Which -- according to documentation -- means that the bool query has been executed in query context:

> **SHOULD**: The clause (query) should appear in the matching document. If the bool query is in a query context and has a must or filter clause then a document will match the bool query even if none of the should queries match. In this case these clauses are only used to influence the score. If the bool query is a filter context or has neither must or filter then at least one of the should queries must match a document for it to match the bool query. This behavior may be explicitly controlled by settings the minimum_should_match parameter.

Since the bool query has a *must* clause (`{"match": {"tags": "a"}}`) that matches both documents, the behaviour of *should* depends on execution context. According to the quoted text in *filter* context there's an implicit `minimum_should_match=1`. The fact that both documents matched suggests that it got executed in query context though. But that seems counter intuitive to:

> Filter context is in effect whenever a query clause is passed to a filter parameter, such as the filter or must_not parameters in the bool query, the filter parameter in the constant_score query, or the filter aggregation.

Also, [this fragment from the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html) says:

> The scores produced by the filtering query of each function do not matter.

Which would suggest filter context.

### Final words

I'm not 100% sure whether it's really an implementation bug, documentation bug or me not understanding the docs and would be happy if someone could help me figure that out.

### PR DESCRIPTION
**Pull request by jimczi**
When building a query Lucene distinguishes two cases, queries that require to produce a score and queries that only need to match. We cloned this mechanism in the QueryBuilders in order to be able to produce different queries based on whether they need to produce a score or not. However the only case in es that require this distinction is the BoolQueryBuilder that sets a different minimum_should_match when a `bool` query is built in a filter context. This behavior doesn't seem right because it makes the matching of `should` clauses different when the score is not required.

Closes #35293

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java [hunk 1]
```diff
@@ -384,30 +384,14 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
             return new MatchAllDocsQuery();
         }
 
-        final String minimumShouldMatch;
-        if (context.isFilter() && this.minimumShouldMatch == null && shouldClauses.size() > 0) {
-            minimumShouldMatch = "1";
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java]: This is the part that confuses me. Why do we need to set `minimum_should_match` in this case ? I cannot think of one example where this fixes something ? @martijnvg can you confirm ?

**Comment by javanna** [comment]
heya @jimczi I was wondering, where is this change breaking?

**Review by mayya-sharipova — APPROVED** [review]
I wonder if also need to modify  or completely remove `query_filter_context.asciidoc` https://github.com/elastic/elasticsearch/blob/master/docs/reference/query-dsl/query_filter_context.asciidoc 

Because as I understand with your PR, we will not have query and filter contexts any more. And also what this document says about caching doesn't seem to be right (as I understood we don't cache based on query and filter context, but based on the query type)

**Review by jimczi — COMMENTED** [review]
> I wonder if also need to modify or completely remove query_filter_context.asciidoc https://github.com/elastic/elasticsearch/blob/master/docs/reference/query-dsl/query_filter_context.asciidoc

It's still relevant. The explanation implies that the query is sorted by relevancy so overall score of the query is taken into account. In this case the differenciation between a `filter` and a `should` clause is important. Same with the `constant_score` query. 

I think the main question here is whether we want to make this breaking change in 7 or not.
In such case we'd also need to add deprecation warnings in 6 for queries with a mixed of should and required clauses in a filter context. If we feel that it's too late for 7 we could only deprecate now with the goal to remove it in 8.

**Review by jpountz — APPROVED** [review]
+1 this distinction was mostly useful for the `terms` query initially, which was parsed as a BooleanQuery when scoring was enabled and TermsQuery in a filter context. We should remove it now, I agree that the current behavior on `bool` queries is quite confusing.

**Comment by jimczi** [comment]
Thanks @jpountz , we discussed this some time ago and I forgot to update the pr. We've agreed that this change should be done in ES 7 and we'll deprecate the behavior in 6x. I'll update the pr with the missing breaking change and prepare the deprecation.

**Comment by matthuhiggins** [comment]
With this change, is this the shortest way to perform an unscored "OR"?
```
{
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "should": [
            { "term": {"city": "Portland"}},
            { "term": {"state": "Washington"}}
          ],
         "minimum_should_match": 1
        }
      }
    }
  }
}
```

**Comment by telendt** [comment]
@matthuhiggins This is a proper way of performing "unscored disjunction (OR)". It's not the shortest way though as `minimum_should_match` defaults to 1 if there are only should clauses (this is a default Lucene behavior). It still ok to specify `minimum_should_match` explicitly (it's also how I prefer to do it).

**Comment by matthuhiggins** [comment]
Got it, thanks. The release notes and deprecation warning leads one to think  "minimum_should_match" always defaults to 0, even in the case when there are only "should" clauses.

**Pull request merged**